---
title: "makeExp2"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## R Markdown

### Creates mixture data from scRNA-seq data and saves expression matrix, trueH, trueW, unmixed to exp2.RData

#### Load parsed scRNA-seq data 
```{r}
#load("~/rnaGinesis/data/muraro.seurat.RData")
load("~/rnaGinesis/data/muraro.alisa.RData")
```

#### Creates mu (avg W) with cell-type gene expression profiles
```{r}
mu   = makeCentroid(data)

# Subsets mu to contain gene expression profiles for 3 most abundant cell types
mu   = subsetCentroid(data, mu, 3)
```


##### Creates H with cell-type proportions in each sample, dimensions cell types by samples 
```{r}
# Creates H with 3 cell types, Samplesize = 50
Samplesize = 50
d.params   = c("alpha"  = 0.5,
               "beta"   = 0.2,
               "duct" = 0.3)
trueH      = sim_H(S = Samplesize,
           d.params = d.params)
``` 

#### Obtains mixture expression matrix data and list of individual Ws for each sample
```{r}
total  = 500 # Number of cells want to sample out of population of all cells in data
result = make_W(data, 
                 trueH, 
                 total)
ex     = result[[1]]
trueW  = result[[2]]
```

#### Saves expression matrix, trueW, trueH, unmixed signatures (mu)
```{r}
experiment2         = list()
experiment2$ex      = ex
experiment2$trueW   = trueW
experiment2$trueH   = trueH
experiment2$unmixed = mu
```

#### Saves experiment2 data to .RData file
```{r}
save(experiment2, file = "~/rnaGinesis/data/alisa_exp2.RData")
```